src/lib.rs

```diff
@@ -1,55 +1,97 @@
-//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Consensus] struct
-//! requires access to an [Environment] object which black-boxes the network layer and gives
-//! appropriate access to the set of available blocks that we need to make consensus on.
+//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Member] struct
+//! requires access to a network layer, a cryptographic primitive, and a data provided that
+//! gives appropriate access to the set of available data that we need to make consensus on.
 
 use codec::{Decode, Encode};
-use futures::{Future, Sink, Stream};
-use log::{debug, error};
+use futures::Future;
 use std::{
     fmt::{Debug, Display},
-    hash::Hash,
-};
-use tokio::{
-    sync::{mpsc, oneshot},
-    time::Duration,
+    hash::Hash as StdHash,
 };
+use tokio::sync::mpsc;
 
-use crate::{creator::Creator, extender::Extender, syncer::Syncer, terminal::Terminal};
+use crate::nodes::{NodeCount, NodeIndex, NodeMap};
 
-pub use crate::nodes::{NodeCount, NodeIndex, NodeMap};
+pub use member::{Config, Member};
 
+mod bft;
+mod consensus;
 mod creator;
 mod extender;
+mod member;
 pub mod nodes;
 mod syncer;
 mod terminal;
 mod testing;
+mod units;
+
+pub trait DataProvider<Data> {
+    type Error: Debug;
+    fn get_data(&self) -> Data;
+    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
```

Comment:
I would call this whole thing DataHandler, so that it is both a Provider and a Sink. For batches, there is some low-level property of consensus that naturally results in producing batches instead of single "Data"s.  I don't have anything against batching, but I don't see whether it helps in anything, so an API which sends single pieces of Data is also fine.

Reply:
How about renaming DataProvider to DataIO?

Reply:
Changed.